src/cli/java/org/commcare/util/screen/MultiSelectEntityScreen.java

```diff
@@ -0,0 +1,97 @@
+import org.commcare.modern.session.SessionWrapper;
+import org.commcare.session.CommCareSession;
+import org.commcare.suite.model.MultiSelectEntityDatum;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.core.util.PropertyUtils;
+
+import java.sql.SQLException;
+
+/**
+ * Variation of EntityScreen to allow for selection of multiple entities at once
+ */
+public class MultiSelectEntityScreen extends EntityScreen {
+
+    public static final CharSequence USE_SELECTED_VALUES = "use_selected_values";
+    private int maxSelectValue = -1;
+
+    private final EntitiesSelectionCache entitiesSelectionCache;
+    private String storageReferenceId;
+
+    public MultiSelectEntityScreen(boolean handleCaseIndex, boolean full,
+            SessionWrapper session,
+            EntitiesSelectionCache entitiesSelectionCache)
+            throws CommCareSessionException {
+        super(handleCaseIndex, full, session);
+        this.entitiesSelectionCache = entitiesSelectionCache;
+    }
+
+    public void setSelectedEntities(String input, String[] selectedValues)
+            throws CommCareSessionException {
+        if (input.contentEquals(USE_SELECTED_VALUES)) {
+            try {
+                processSelectedValues(selectedValues);
+            } catch (SQLException throwables) {
+                throw new CommCareSessionException(
+                        "An error occurred trying to process selections on this screen. " +
+                                " If this error persists please report a bug to CommCareHQ.", throwables);
+            }
+        } else {
+            String[] cachedSelection = entitiesSelectionCache.read(input);
+            if (cachedSelection == null) {
+                throw new CommCareSessionException(
+                        "Could not make selection with reference id " + input + " on this screen. " +
+                                " If this error persists please report a bug to CommCareHQ.");
+            }
+            storageReferenceId = input;
+        }
+    }
+
+    private void processSelectedValues(String[] selectedValues)
+            throws SQLException, CommCareSessionException {
+        if (selectedValues != null) {
+            String[] evaluatedValues = new String[selectedValues.length];
+            for (int i = 0; i < selectedValues.length; i++) {
+                TreeReference currentReference = getEntityReference(selectedValues[i]);
+                if (currentReference == null) {
+                    throw new CommCareSessionException(
+                            "Could not select case " + selectedValues[i] + " on this screen. " +
+                                    " If this error persists please report a bug to CommCareHQ.");
+                }
+                evaluatedValues[i] = getReturnValueFromSelection(currentReference);
+            }
+
+            String guid = PropertyUtils.genGUID(10);
+            entitiesSelectionCache.cache(guid, evaluatedValues);
```

Comment:
when does this get cleared?

Reply:
Ahh good point, I didn't think about it. I will look into it now.